File: Python/data_structures/11_valid_binary_search_tree.py
import sys
from collections import namedtuple

Node = namedtuple("Node", ["key", "left", "right"])

class Stack(object):
    def __init__(self):
        self._stack = []

    @property
    def elements(self):
        return self._stack

    @property
    def is_empty(self):
        if self._stack:
            return False
        return True

    def top(self):
        return self._stack[-1]

    def push(self, elem):
        self._stack.append(elem)

    def pop(self):
        elem = self._stack.pop(-1)
        return elem


class BinaryTree(object):
    def __init__(self, node_list):
        self._tree = node_list
        self._root_index = 0

    def inorder_traversal(self):
        keys = []
        stack = Stack()
        node_index = self._root_index
        while True:
            if node_index != -1:
                node = self._tree[node_index]
                stack.push(node)
                node_index = node.left
            elif not stack.is_empty:
                node = stack.pop()
                keys.append(node.key)
                node_index = node.right
            else:
                return keys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def read_node_information(reader, nr_nodes):
    return tuple(Node(*convert_to_intlist(next(reader))) for _ in range(nr_nodes))


def is_sorted(lst):
    return all(elem1 <= elem2 for elem1, elem2 in zip(lst, lst[1:]))


def is_valid_binary_search_tree(tree, number_of_nodes):
    if number_of_nodes <= 1:
        return True
    keys = tree.inorder_traversal()
    return is_sorted(keys)


def main():
    reader = sys.stdin
    number_of_nodes = int(next(reader))
    nodes = read_node_information(reader, number_of_nodes)
    tree = BinaryTree(nodes)
    result = is_valid_binary_search_tree(tree, number_of_nodes)
    if result:
        print("CORRECT")
    else:
        print("INCORRECT")


if __name__ == "__main__":
    main()

File: Julia/data_structures/11-valid-binary-search-tree.jl
struct Node
    key::Int
    left::Int
    right::Int
end


struct BinaryTree
    root::Int
    nodes::Vector{Node}

    BinaryTree(nodes::Vector{Node}) = new(0, nodes)
end


function inorder(tree::BinaryTree)
    keys = Int[]
    stack = Node[]
    nodeindex = tree.root
    while true
        if nodeindex != -1
            node = tree.nodes[nodeindex+1]
            push!(stack, node)
            nodeindex = node.left
        elseif !isempty(stack)
            node = pop!(stack)
            push!(keys, node.key)
            nodeindex = node.right
        else
            return keys
        end
    end
end


converttointlist(line) = map(x -> parse(Int, x), split(line))


function readnodes(nrnodes)
    nodes = Node[]
    for _ in 1:nrnodes
        line = readline() |> converttointlist
        push!(nodes, Node(line...))
    end
    nodes
end


function validsearchtree(tree, nrnodes)
    if nrnodes <= 1
        true
    else
        keys = inorder(tree)
        issorted(keys)
    end
end


function main()
    nrnodes = parse(Int, readline())
    nodes = readnodes(nrnodes)
    tree = BinaryTree(nodes)
    println(validsearchtree(tree, nrnodes) ? "CORRECT" : "INCORRECT")
end


main()
